File: 0572-subtree-of-another-tree/0572-subtree-of-another-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isSubtree(TreeNode root, TreeNode subRoot) {
        if (root == null && subRoot == null) return true;
        if (root == null || subRoot == null) return false;

        return same(root, subRoot) || isSubtree(root.left, subRoot) || isSubtree(root.right, subRoot);
    }
    private boolean same(TreeNode lhs, TreeNode rhs) {
        if (lhs == null && rhs == null) return true;
        if (lhs == null || rhs == null) return false;

        return lhs.val == rhs.val && same(lhs.left, rhs.left) && same(lhs.right, rhs.right);
    }
}

File: 0572-subtree-of-another-tree/0572-subtree-of-another-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if not root and not subRoot:
            return True
        if bool(root) ^ bool(subRoot):
            return False
        
        def equals(lhs, rhs):
            if bool(lhs) ^ bool(rhs):
                return False
            if not lhs and not rhs:
                return True
            return lhs.val == rhs.val and equals(lhs.left, rhs.left) and equals(lhs.right, rhs.right)
        
        return equals(root, subRoot) or \
            self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot)
        
